docs/types/types.md

```diff
@@ -0,0 +1,299 @@
+Book "Jane Austen" "Pride and Prejudice" 1832 : Book Text Text Int
+```
+
+The eagle-eyed among you might note that author_type and author are different field names, so this would actually elaborate out to
+
+```
+type Book.type author title year
+  type Book (author : author) (title : title) (year : year)
+```
+
+where we are careful with lexical scopes to distinguish between the term level author and the type level author, enabling this pun.
+
+In the List example
+
+```
+type List a
+  type Nil
+  type Cons (head : a) (tail : List a)
+```
+
+this elaborates each of these type declarations and finds a couple of puns:
+
+```
+type List.type a = List a
+type Nil.type = Nil
+```
+
+and one definition that emerges as not a pun:
+
+```
+type Cons.type a
+  type Cons (head : a) (tail : List a)
+```
+
+# static methods and type formation
+
+List serves double duty in the above. It acts as the type constructor for lists, so that `List Int` is a well formed `Type`, but it also
+acts as the container holding its static methods.
+
+There are two forms of static methods that we are interested in in the long term. We'll classify them as 'normal' static methods and 'explicit' static methods. For now we're going to support 'normal' static methods as they are needed for 'from' conversions, but there is an opportunity to allow more with explicit statics later on for some uses involving meta-programming.
+
+```
+type Vector a
+  type MkVec (unsafe_to_array: Array a) 
+  static from (that : List x): Vector x = MkVec (Array.from that)
+```
+
+this internally elaborates to
+
+```
+type Vector a
+  type MkVec.type a
+    type MkVec (unsafe_to_array: Array a)
+  ...
+```
+
+with the obvious subtyping relationship
+
+```
+MkVec.type a <: Vector a
+```
+
+Similarly to the pun story above, when you have a single `type` statement contained in a parent `type` statement and the shape aligns, we
+should be willing to allow punning between them, eliminating the distinct `MkVec.type` here. This results a nice "obvious" usage pattern, where
+we aren't distinguishing between several vector types. The type of the module it lives in the type constructor, and the type of its individual constuctor that is limited to only holding arrays.
+
+Use `Vector` itself as the place where the statics live allows for a convenient usage pattern:
+
+```
+Vector.from [1,2,3] : Vector Int
+```
+
+where 'Vector' is playing double duty as a type constructor and as the bag of static methods.
+
+```
+Vec ([1,2,3].to Array) : Vector Int
+```
+
+Since static methods live in the topmost atom _constructor_, this implies a constraint on type statements that contain statics. Namely that it has to pun with its own type. This translates to having no distinct type variables for each of its arguments. We may choose to relax this in the future, but by being overstrict right now we leave room for evolving the language spec, rather than risk users depending on bad behaviors we'd have to change later.
```

Comment:
```suggestion
Since static methods live in the topmost atom _constructor_, this implies a constraint on type statements that contain statics. Namely that it has to pun with its own type. This translates to having no distinct type variables for each of its arguments. We may choose to relax this in the future, but by being overstrict right now, we leave room for evolving the language spec rather than risk users depending on bad behaviors we'd have to change later.
```